pyro/poutine/trace.py

```diff
@@ -20,105 +19,265 @@ def _warn_if_nan(name, value):
+
+
+def is_non_reparameterized(site):
+    return site["type"] == "sample" and \
+        not site["is_observed"] and \
+        not getattr(site["fn"], "reparameterized", False)
 
 
 class Trace(object):
     """
-    Execution trace data structure
+    Execution trace data structure.
     """
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, graph_type="flat"):
         """
         :param string graph_type: string specifying the kind of trace graph to construct
 
-        Constructor. Currently identical to networkx.``DiGraph(\*args, \**kwargs)``,
-        except for storing the graph_type attribute
+        Constructor. Stores graph_type attribute and creates data dictionaries.
         """
-        self._graph = DiGraph(*args, **kwargs)
-        graph_type = kwargs.pop("graph_type", "flat")
         assert graph_type in ("flat", "dense"), \
             "{} not a valid graph type".format(graph_type)
         self.graph_type = graph_type
-        super(Trace, self).__init__(*args, **kwargs)
+        # metadata
+        self.graph = collections.OrderedDict()
+        self.nodes = collections.OrderedDict()
+        self.edges = collections.OrderedDict()
 
-    def __del__(self):
-        # Work around cyclic reference bugs in networkx.DiGraph
-        # See https://github.com/uber/pyro/issues/798
-        self._graph.__dict__.clear()
+    def __contains__(self, x):
+        return x in self.nodes
 
-    @property
-    def nodes(self):
-        return self._graph.nodes
+    def __getitem__(self, name):
+        return self.edges[name]
 
-    @property
-    def edges(self):
-        return self._graph.edges
-
-    @property
-    def graph(self):
-        return self._graph.graph
-
-    @property
-    def remove_node(self):
-        return self._graph.remove_node
-
-    @property
-    def add_edge(self):
-        return self._graph.add_edge
-
-    @property
-    def is_directed(self):
-        return self._graph.is_directed
+    def __len__(self):
+        return len(self.nodes)
 
-    @property
-    def in_degree(self):
-        return self._graph.in_degree
+    def __iter__(self):
+        return iter(self.nodes)
 
-    @property
-    def successors(self):
-        return self._graph.successors
+    def successors(self, site_name):
+        """
+        :param string site_name: name of site to get successors of
 
-    def __contains__(self, site_name):
-        return site_name in self._graph
+        Returns an iterator over site names whose parent is site_name
+        """
+        return iter(self.edges[site_name])
 
-    def __iter__(self):
-        return iter(self._graph)
+    def descendants(self, site_name):
+        """
+        :param string site_name: name of site to get descendants of
 
-    def __len__(self):
-        return len(self._graph)
+        Returns a list of all sites downstream of site_name
+        """
+        desc = set()
+        for succ in self.edges[site_name]:
```

Comment:
This implementation is exponential in number of nodes. Why not make it linear via depth-first search?
```py
frontier = set([site_name])
desc = set()
while frontier:
    node = frontier.pop()
    if node not in desc:
        desc.add(node)
        frontier.update(set(self.edges[node]) - desc)
desc.remove(site_name)
return list(sorted(desc))
```